### PR TITLE
Correction on switch to bundled JS or not + add parameter to be in de…

### PR DESCRIPTION
**Pull request by mbaroukh**
Hi.

Why this pull request :
On Android, we can make many variant.
For each variant, you can specify if js is bundled into the apk.
But you can't specify if JS will be in developpement mode or production mode because the selection on dev mode or prod mode is made with :
`def devEnabled = !targetName.toLowerCase().contains("release")`

So this pull request only add a configuration option to allow it. Suppose I have a build named "Rec7", I can :
`project.ext.react = [ 
        bundleInRec7: true,
        devModeForRec7: false,
]`

By the way there is a bug in allowing or not to bundle js in apk.
The selection is made with this code in local-cli/generator-android/templates/src/app/build.gradle :
`enabled config."bundleIn${targetName}" ||
                    config."bundleIn${buildTypeName.capitalize()}" ?:
                            targetName.toLowerCase().contains("release")
`
It works well for all varant except if there is "release" in it : 
in this case, if you use "bundleInRelease: false", the evaluation will go to next part and finally, the only evaluation that will be used is "targetName.toLowerCase().contains("release")".
I understand that most times, we wan't to bundle for release, but if the otption exists, it should work as expected.

So I just changed for 
`
config."bundleIn${targetName}"!=null?config."bundleIn${targetName}":
                    config."bundleIn${buildTypeName.capitalize()}"!=null?config."bundleIn${buildTypeName.capitalize()}":
                            targetName.toLowerCase().contains("release")
`
that will continue to next par if config parameter is null and only if it's null. If it's defined, even if false, it will be used.

Notes:
- What could be confusing is that devMode parameter will only be valid when js is bundled. I'll try to see if we can use this parameter while using http packager
- Also, I think there is an issue when changing devMode : There must be some caching and bundle did not always occur. And if bundle is not done, the bundled JS will not toggle from/to dev/prod.

Thanks.

Mike

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @philikon and @kgritesh to be potential reviewers.

**Comment by facebook-github-bot-5** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-3** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

File: local-cli/generator-android/templates/src/app/build.gradle [hunk 1]
```diff
@@ -24,6 +24,14 @@ import com.android.build.OutputFile
  *   // whether to bundle JS and assets in release mode
  *   bundleInRelease: true,
  *
+ *   // whether to have JS in dev mode for a variant (with warnings)
+ *   devModeFor[VARIANT]: true|false,
+ *   (default : true only if variant contains "release")
+ *   Note: this param only apply if JS files are bundled.
+ *
+ *   // whether to bundle JS and assets in release mode
+ *   bundleInRelease: true,
```
Comment (mkonicek) [inline local-cli/generator-android/templates/src/app/build.gradle]: Looks like this is now present twice?

     // whether to bundle JS and assets in release mode
     bundleInRelease: true,

**Comment by mkonicek** [comment]
> But you can't specify if JS will be in development mode or production mode because the selection on dev mode or prod mode is made with :
def devEnabled = !targetName.toLowerCase().contains("release")

I'm curious - why have release JS bundle in a dev Android build, or debug JS bundle in a release build?

**Comment by mbaroukh** [comment]
> I'm curious - why have release JS bundle in a dev Android build, or debug JS bundle in a release build?

Me too, but that is not the subject :).
1/ if there is an option to enable or not devMode, it should work what ever the variant is. Actually, it don't work for release.
2/ if you wan't another build than release in production mode, you can't. (this was my problem : if I deliver a test app to my client, I don't wan't it to be in dev mode because of all the warnings. And yes : I should have no warning and I could call the variant "pre-release" ... )

**Comment by facebook-github-bot-4** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @philikon as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by facebook-github-bot-6** [comment]
@mbaroukh updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9602/files/8f472b8a70b84abae6f7820aa93101b8af612914..3882070fde3f7f57aa754d84311be4933e54e7bd)

**Comment by facebook-github-bot-3** [comment]
@mbaroukh updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9602/files/3882070fde3f7f57aa754d84311be4933e54e7bd..53ff9f0b404a31691c2fad2069324e7c0725eb5b)

**Comment by facebook-github-bot-6** [comment]
@mbaroukh updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9602/files/53ff9f0b404a31691c2fad2069324e7c0725eb5b..f84a81884129f7bb3e9aad4001346d2a9bcc82e9)

**Comment by mbaroukh** [comment]
Don't know what going on with tests, but I just realized that I could have run tests on CircleCI before submitting this PR. Sorry. Ill try to debug on CircleCI.

**Comment by hramos** [comment]
It's still not clear to me why this is needed.

**Comment by hramos** [comment]
I am not sure I understand why this change is necessary. @mkonicek thoughts?

**Comment by mbaroukh** [comment]
This change is needed if you wan't to disable dev mode for other builds than release. 
At least, that's just another option and if nothing changed on default config, it will works exactly like actually. 
But before giving a release build to my customer, I give him a "pre-release" build that is called "recette" in french. There is not the word release in it and so with the actual react.gradle file, I'm not able to give them a build that is not devMode enabled.

the actual react.gradle file use the variant name to configure if this is a dev or not build. 
but when configuring variant, you have the option to configure it with "debuggable true/false" so that's not consistent to have the option to configure a not debuggable build on every variant but not being able to say that is not debuggable build while display react warnings.

I see that javascript world usually have on dev or production build, nothing else.
Maybe it's the same on ios, I don't know.
But on android, you can have lots of variant.

For the ci that failed, I don't understand why. I tried to log on travis and make some tests but without success. I used this modified react.gradle file every day at work without any issue.

**Comment by cong876** [comment]
@mbaroukh 
while I build the android, I find the same problem with you.
Finally I solve the problem by config like this
`
project.ext.react = [
    bundleIn${productFlavor}${buildType}: true,
    bundleIn${buildType}: true
]
`

What cause this problem is that I define the new productFlavors.

**Comment by lacker** [comment]
Hey, thanks for submitting this pull request! I think the net conclusion here is that we would prefer that this functionality not go into React Native. It seems like a decent workaround that you can have this flag in a fork but we keep it out of the regular build. So I am going to close this pull request, but thanks again for bringing up the issue.

**Pull request closed without merging**